### PR TITLE
Add solution for BinarySearchTree

### PR DESCRIPTION
**Pull request by Showndarya**
Solution for the BinaryTree problem in Session8. 
Answer for question in BinarySearchTree.java in comments.
Comments for each function also included.

I observed that the solution for LinkedList was present already. So, didn't add that.

### CODE REVIEW

File: Session08/binarytree/solutions/BinarySearchTree.java [hunk 1]
```diff
@@ -0,0 +1,123 @@
+// This class fails to compile, can you figure out why, and fix it?
+// This is beacuse BinaryTree has a constructor which is not default. Hence, when we extend this class and try to call a no-argument constructor, it'll give a compile-time error. so, we call it by passing the parameters (to instantiate the root node) in super() called inside the constructor of BinarySearchTree
+
+class BinarySearchTree extends BinaryTree {
+
+    private int flag = 0;
+    
+    public BinarySearchTree()
+    {
+        super(0);
```
Comment (havanagrawal) [inline Session08/binarytree/solutions/BinarySearchTree.java]: Why is the root randomly 0? It will be extremely unexpected if a user constructs an "empty" binary search tree, but minValue or maxValue will return 0.

File: Session08/binarytree/solutions/BinarySearchTree.java [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    }
+    
+    // insert node
+	@Override
+	public boolean insert(int value) {
+		super.root = insert(super.root, value);
```
Comment (havanagrawal) [inline Session08/binarytree/solutions/BinarySearchTree.java]: What if the node already exists? This function always returns true regardless of whether the insert "fails" or succeeds.

File: Session08/binarytree/solutions/BinarySearchTree.java [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+		return true;
+	}
+
+    private TreeNode insert(TreeNode root, int data)
+    {
+        // if current node is null, insert node with data
+        if (root == null) root = new TreeNode(data);
```
Comment (havanagrawal) [inline Session08/binarytree/solutions/BinarySearchTree.java]: If both your if and else blocks have a single line, this formatting is fine, like the way you have used below.

However, in this case, please adhere to the repo's coding style of

```java
if (<condition>) {
}
else {
}
```

Curly braces are good. Always use them, even when they are not needed.

File: Session08/binarytree/solutions/BinarySearchTree.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+            else root.right = insert(root.right, data);
+        }
+        return root;
+    }
+    
+    // delete node
+	@Override
+	public boolean delete(int value) {
+		super.root = deleteRec(super.root, value);
+		if(flag==1) {flag=0;return false;}
```
Comment (havanagrawal) [inline Session08/binarytree/solutions/BinarySearchTree.java]: Please format as 

```java
if (flag == 1) {
    flag = 0;
    return false;
}
```

Note the spaces

File: Session08/binarytree/solutions/BinarySearchTree.java [hunk 5]
```diff
@@ -0,0 +1,123 @@
+// This class fails to compile, can you figure out why, and fix it?
+// This is beacuse BinaryTree has a constructor which is not default. Hence, when we extend this class and try to call a no-argument constructor, it'll give a compile-time error. so, we call it by passing the parameters (to instantiate the root node) in super() called inside the constructor of BinarySearchTree
+
+class BinarySearchTree extends BinaryTree {
+
+    private int flag = 0;
```
Comment (havanagrawal) [inline Session08/binarytree/solutions/BinarySearchTree.java]: Any reason why this is not a `boolean`? Also, why is it a member variable? What does "BinarySearchTree.flag" mean for the tree as a whole?

File: Session08/binarytree/solutions/BinarySearchTree.java [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    
+    // delete node
+	@Override
+	public boolean delete(int value) {
+		super.root = deleteRec(super.root, value);
+		if(flag==1) {flag=0;return false;}
+		else return true;
+	}
+	
+	TreeNode deleteRec(TreeNode root, int key)
```
Comment (havanagrawal) [inline Session08/binarytree/solutions/BinarySearchTree.java]: Please follow code style of opening the curly brace on the same line.

File: Session08/binarytree/solutions/BinarySearchTree.java [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+        {
+            // traverse to left sub-tree if value to be inserted is lesser than value in current node
+            if (data <= node.getData()) root.left = insert(root.left, data);
+            // traverse to right sub-tree if value to be inserted is greater than value in current node
+            else root.right = insert(root.right, data);
+        }
+        return root;
+    }
+    
+    // delete node
```
Comment (havanagrawal) [inline Session08/binarytree/solutions/BinarySearchTree.java]: Is this comment really necessary? The function name is `delete`. Self-documenting code.

File: Session08/binarytree/solutions/BinarySearchTree.java [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+class BinarySearchTree extends BinaryTree {
+
+    private int flag = 0;
+    
+    public BinarySearchTree()
+    {
+        super(0);
+    }
+    
+    // insert node
```
Comment (havanagrawal) [inline Session08/binarytree/solutions/BinarySearchTree.java]: Is this comment really necessary? The function name is `insert`. Self-documenting code.

File: Session08/binarytree/solutions/BinarySearchTree.java [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+            // if node has two children, get the lowest inorder successor from the right subtree 
+            root.data = minValue(root.right); 
+            // Delete the inorder successor
+            root.right = deleteRec(root.right, root.data);
+        } 
+        return root;
+    }
+    
+    // find the lowest value in the tree by reaching the leftmost leaf node
+    int minValue(TreeNode root)
```
Comment (havanagrawal) [inline Session08/binarytree/solutions/BinarySearchTree.java]: What happens when the tree is empty?

File: Session08/binarytree/solutions/BinarySearchTree.java [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+        int minv = root.data;
+        while (root.left != null)
+        {
+            minv = root.left.data;
+            root = root.left;
+        }
+        return minv;
+    }
+	
+	// check if node exists
```
Comment (havanagrawal) [inline Session08/binarytree/solutions/BinarySearchTree.java]: Is this comment really necessary? The function name is `exists`. Self-documenting code.

File: Session08/binarytree/solutions/BinaryTree.java [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+	   it is useful because the child classes can use it,
+	   by using super(), to instantiate the root node
+	 */
+	public BinaryTree(int rootValue) {
+		this.root = new TreeNode(rootValue);
+	}
+
+	public void inOrderTraversal() {
+		//TODO
+		inorderRec(root);
```
Comment (havanagrawal) [inline Session08/binarytree/solutions/BinaryTree.java]: I am assuming that `inOrderRec()` stands for `inOrderRecursive`. However, there is no way for me to know this, since it's not a convention. Please name it `inOrderRecursive`

File: Session08/binarytree/solutions/TreeNode.java [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+	int data;
+
+	public TreeNode(int data) {
+		this.data = data;
+		this.left = null;
+		this.right = null;
+	}
+	
+	// set left node 
+	public void setLeft(TreeNode n)
```
Comment (havanagrawal) [inline Session08/binarytree/solutions/TreeNode.java]: This is really bad! We do not want setters on a `TreeNode`. Imagine what happens if I insert a TreeNode into a BinarySearchTree, and then set its value to be something else. Now you've violated the BinarySearchTree invariant. 

Additionally, member variables without an explicit access modifier are `default`, and have package level access. Thus we don't really need getters either. 

(In the ideal case, the `TreeNode` class would be immutable.)

File: Session08/binarytree/solutions/BinarySearchTree.java [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+    {
+        // if tree is empty, set flag to 1 and return root
+        if (root == null)  {flag=1;return root;}
+ 
+        // traverse to left sub-tree if value to be inserted is lesser than value in current node
+        if (key < root.data) root.left = deleteRec(root.left, key);
+        // traverse to right sub-tree if value to be inserted is greater than value in current node
+        else if (key > root.data) root.right = deleteRec(root.right, key);
+ 
+        // if value is same as current value, delte it
```
Comment (havanagrawal) [inline Session08/binarytree/solutions/BinarySearchTree.java]: typo in comment, should be "delete"

File: Session08/binarytree/solutions/BinarySearchTree.java [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+		return true;
+	}
+
+    private TreeNode insert(TreeNode root, int data)
+    {
+        // if current node is null, insert node with data
+        if (root == null) root = new TreeNode(data);
+        else
+        {
+            // traverse to left sub-tree if value to be inserted is lesser than value in current node
```
Comment (havanagrawal) [inline Session08/binarytree/solutions/BinarySearchTree.java]: Your comment says "lesser than", but your code says "lesser than _or equal to"_. Decide which one is incorrect.

**Comment by havanagrawal** [comment]
**tl;dr**
Hi @Showndarya , thank you for the PR. I appreciate your initiative, and it is an excellent attempt. Please see comments, and update your PR.

**The long version**
I see that you're a student in the C2C 2018 cohort, and once again, I admire your willingness to learn. Please do note that the above comments are something you would have learned before we reached this topic. We focus a lot on good design.

Some broad improvements you can make:

1. Follow the code-style of the project. Look at how the code is structured in the rest of the repo, and try to mimic it. Before you contribute to any project, it is important that you follow their code-style, rather than your personal choice.
2. Don't use global "flags".
3. Don't use an IDE (at this stage). It inserts boilerplate for you, and more often than not, you don't need it (at least not for this course). You should think long and hard before you add a public function to a class.
4. Contrary to popular belief, comments aren't really that great. Robert Martin (Uncle Bob) has a fantastic one-liner on this, (I paraphrase) "It is a mistake to think that a programmer who failed to express his thought in code will be able to do it in English." Avoid comments. Put them only where absolutely necessary, and then think if you can restructure your code to make the idea clearer.

If you want, go ahead and make the above changes, and I'll review again. 

Cheers and happy learning 😄

File: Session08/binarytree/solutions/BinarySearchTree.java [hunk 15]
```diff
@@ -0,0 +1,123 @@
+// This class fails to compile, can you figure out why, and fix it?
+// This is beacuse BinaryTree has a constructor which is not default. Hence, when we extend this class and try to call a no-argument constructor, it'll give a compile-time error. so, we call it by passing the parameters (to instantiate the root node) in super() called inside the constructor of BinarySearchTree
```
Comment (havanagrawal) [inline Session08/binarytree/solutions/BinarySearchTree.java]: Please use multiline comments for long comments.

```java
/* line1
 * line2
 */
```

File: Session08/binarytree/solutions/BinarySearchTree.java [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+    // lowest common ancestor
+	@Override
+	public TreeNode lowestCommonAncestor(int value1, int value2) {
+		
+		TreeNode tn = lca(super.root, value1,value2);
+		if(flag==1) return null;
+		else return tn;
+	}
+	
+	// the node value present in between a and b is the lca. Assuming, a<b
```
Comment (havanagrawal) [inline Session08/binarytree/solutions/BinarySearchTree.java]: Why assume that a < b? Why not ensure it?

File: Session08/binarytree/solutions/BinarySearchTree.java [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+	public TreeNode lca(TreeNode root, int a, int b)
```
Comment (havanagrawal) [inline Session08/binarytree/solutions/BinarySearchTree.java]: Why does the class have two public methods that do the exact same thing?

File: Session08/binarytree/solutions/BinarySearchTree.java [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+	{
+	    if(root == null) {flag=1;return null;}
+	    
+	    if(root.data>b && root.data>a) lca(root.left,a,b);
```
Comment (havanagrawal) [inline Session08/binarytree/solutions/BinarySearchTree.java]: If you have assumed that a < b, is it really necessary to check if `root.data` is greater than both of them?

File: Session08/binarytree/solutions/BinarySearchTree.java [hunk 19]
```diff
@@ -0,0 +1,123 @@
...
+	    else if(root.data<b && root.data<a) lca(root.right,a,b);
```
Comment (havanagrawal) [inline Session08/binarytree/solutions/BinarySearchTree.java]: If you have assumed that a < b, is it really necessary to check if `root.data` is lesser than both of them?

**Comment by Showndarya** [comment]
Thank you for the review.

I have taken all the improvements and suggestions into consideration and have come up with a more refined solution.

Let me know if more changes are required.
